### PR TITLE
feat: Improve support for unneeded parentheses removal

### LINKED ISSUE #1812: Extra brackets removal fixer proposal
For example:

``` php
$stuff = ($this->baz) ? $this->foo() : $this->bar();
```

We don't need the brackets around `$this->baz`.

### PR DESCRIPTION
**Pull request by Wirone**
Fixes #1812

I wanted to verify cases mentioned in the issue and started to play around with the fixer. I have few more test cases and part of them are already covered by fixer. I want to discuss others, hence draft PR.

### CODE REVIEW

**Comment by Wirone** [comment]
Hah, I did not run `testFixAll()` locally and it was a mistake 😅. So much regression. Few questions:

- do we want to provide fixes for added cases?
- are these cases added in the valid provider, or should be moved to `provideFixAllCases()`?
- is there some convenient way to determine whole call chain like `$foo->bar->baz['a']->{$b}->c` so it would be possible to determine if there is unneeded `()` wrap?
- are new methods used in `isUselessWrapped()` needed, or such logic should be added to existing ones? None of them seemed good to me, but it's hard to understand whole fixer

**Comment by kubawerlos** [comment]
For handling more cases 👍🏼

For breaking PHP syntax (such as `match`, visible in one failed case) 👎🏼

**Comment by Wirone** [comment]
> For breaking PHP syntax (such as match, visible in one failed case) 👎🏼

That's obvious 🙂. These fails are here because I did not run whole test class locally, only one method where cases were added 🤦‍♂️. Before I look at it more, I just wanted to discuss if added cases are OK.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 1]
```diff
@@ -536,6 +544,41 @@ private function isPreUnaryOperation(Tokens $tokens, int $index): bool
...
+                if ($closeIndex === $tokens->getNextMeaningfulToken($invocationEndIndex)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private function isWrappedAssignment(Tokens $tokens, int $openIndex, int $closeIndex): bool
```
Comment (SpacePossum) [inline src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:576]: :+1:  nice one

File: tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php [hunk 2]
```diff
@@ -340,6 +343,47 @@ function foo() { $a = (yield($x)); }
...
+            ],
+            [
+                '<?php $stuff = (Foo::bar())($baz) ? $this->foo() : $this->bar();',
+            ],
+            [
+                // Even if it would work without parentheses in this case, let's ensure we keep them for readability.
+                // Unparenthesised ternaries were deprecated in 7.4 and cause fatal error in 8.0+
+                "<?php
+                echo ((1 == 2 ? 'a' : 3 == 4) ? 'b' : 5 == 6) ? 'c' : 'd';
+                echo 1 == 2 ? 'a' : (3 == 4 ? 'b' : (5 == 6 ? 'c' : 'd'));
```
Comment (SpacePossum) [inline tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php:384]: This is very tricky, as `? :` has `left-associative prior to PHP 8.0.0` , meaning that reliable checking chained `?` and combinations with `??` do not need `(` `)` is pretty hard.
I wouldn't touch ternaries, but if wanted to support, lets get some cases in like:
```php
<?php
$a =
    (($a ?? $b) ?: ($a ? 1 : 2))
    ??
    (((($a ? $b : $c) ?? $a) ?: $d) ?? $y)
?>
```

**Comment by julienfalque** [comment]
Added cases look OK :+1:

**Pull request closed without merging**